# CAPHEINE: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools
- [BioPython](https://biopython.org/)

> Cock PJ, Antao T, Chang JT, Chapman BA, Cox CJ, Dalke A, et al. Biopython: Freely available python tools for Computational Molecular Biology and Bioinformatics. Bioinformatics. 2009 Mar 20;25(11):1422–3. doi:10.1093/bioinformatics/btp163

- [BUSTED](https://doi.org/10.1093/molbev/msv035)

> Murrell B, Weaver S, Smith MD, Wertheim JO, Murrell S, Aylward A, et al. Gene-wide identification of episodic selection. Molecular Biology and Evolution. 2015 Feb 19;32(5):1365–71. doi:10.1093/molbev/msv035 

- [Cawlign](https://github.com/veg/cawlign/tree/main)

- [Contrast-FEL](https://doi.org/10.1093/molbev/msaa263)

> Kosakovsky Pond SL, Wisotsky SR, Escalante A, Magalis BR, Weaver S. Contrast-FEL—a test for differences in selective pressures at individual sites among clades and sets of branches. Molecular Biology and Evolution. 2020 Oct 16;38(3):1184–98. doi:10.1093/molbev/msaa263 

- [FEL](https://doi.org/10.1093/molbev/msi105)

> Kosakovsky Pond SL, Frost SD. Not so different after all: A comparison of methods for detecting amino acid sites under selection. Molecular Biology and Evolution. 2005 Feb 9;22(5):1208–22. doi:10.1093/molbev/msi105 

- [HyPhy](https://doi.org/10.1093/molbev/msz197)

> Kosakovsky Pond SL, Poon AF, Velazquez R, Weaver S, Hepler NL, Murrell B, et al. Hyphy 2.5—a customizable platform for evolutionary hypothesis testing using phylogenies. Molecular Biology and Evolution. 2019 Aug 27;37(1):295–9. doi:10.1093/molbev/msz197 

- [IQ-TREE](https://doi.org/10.1093/molbev/msaa015)

> Minh BQ, Schmidt HA, Chernomor O, Schrempf D, Woodhams MD, von Haeseler A, et al. IQ-tree 2: New models and efficient methods for phylogenetic inference in the genomic era. Molecular Biology and Evolution. 2020 Feb 3;37(5):1530–4. doi:10.1093/molbev/msaa015

- [MEME](https://doi.org/10.1371/journal.pgen.1002764)

> Murrell B, Wertheim JO, Moola S, Weighill T, Scheffler K, Kosakovsky Pond SL. Detecting individual sites subject to episodic diversifying selection. PLoS Genetics. 2012 Jul 12;8(7). doi:10.1371/journal.pgen.1002764 

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

> Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- PRIME

*Citation forthcoming*

- [RELAX](https://doi.org/10.1093/molbev/msu400)

> Wertheim JO, Murrell B, Smith MD, Kosakovsky Pond SL, Scheffler K. Relax: Detecting relaxed selection in a phylogenetic framework. Molecular Biology and Evolution. 2014 Dec 23;32(3):820–32. doi:10.1093/molbev/msu400 

- [SeqKit](https://onlinelibrary.wiley.com/doi/10.1002/imt2.191)

> Shen W, Sipos B, Zhao L. Seqkit2: A Swiss Army knife for sequence and alignment processing. iMeta. 2024 Apr 5;3(3). doi:10.1002/imt2.191 

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

> Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

> Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

> da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

> Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

> Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
